Documentation/ecalength.1
.TH "ecalength" "1" "18\&.04\&.2004" "" "Multimedia software" 
.PP 
.SH "NAME" 
ecalength \- prints audio file lenght and other information
.PP 
.SH "SYNOPSIS" 
\fBecalength\fP [-ahtsfmbcr] FILE1 [FILE2] [FILEn]
.PP 
.SH "DESCRIPTION" 
.PP 
This is ecalength, a few lines of code pompously named so because they  
let one retrieve the length of an audio file from the command line  
using ecasound\'s engine\&.  
.PP 
Limitations:  
.IP "" 
- With files without header information (raw files), ecalength will only work 
correctly if the audio file is at a sampling rate of 44100 hz\&.
(Addressed with the -a switch\&.)
.IP "" 
- It is not foolproof, feeding it with something other than an audio  
file WILL result in ugly things being spewed back\&.  
(A bit better)
.IP "" 
- A thousand more that I haven\'t thought of\&.
.PP 
.SH "OPTIONS" 
.PP 
.IP "-h" 
Prints this usage message\&.  (help)
.IP 
.IP "-a[:]bits,channels,rate" 
Changes the format assumed by default for headerless data\&. (adjust)
.IP 
.IP "-t" 
Prints the summed length of all the files processed\&.  (total)
(Ignored if with -s) 
.IP 
.IP "-s" 
Enables script mode: One info type per file per line\&.   (script)
(Defaults to length in secs\&.) 
.IP 
.IP "-f" 
With -s will return the format string as info, alone it will 
add it to the main display\&.  (format)
.IP 
.IP "-b" 
If -s and -f are enabled with this the info printed will be 
the sample\'s bitwidth\&.  (bits)
.IP 
.IP "-c" 
If -s and -f are enabled with this the info printed will be 
the channel count\&.  (channel count)
.IP 
.IP "-r" 
If -s and -f are enabled with this the info printed will be 
the sampling rate\&.  (rate)
.IP 
.IP "-m" 
Will print human computable time as in main display but in 
batch fashion\&. (minutes) (Only with -s)
.PP 
.SH "SEE ALSO" 
.PP 
ecasound(1), ecatools(1)
.PP 
.SH "AUTHOR" 
.PP 
S\&.Massy, <observer |at| colba\&.net>
.PP 


NEWS
-----------------------------------------------------------------------
** Ecasound - User-visible changes (NEWS)      (last edited: 311001) **
-----------------------------------------------------------------------

About the version numbers... "vX.Y[.Z][devT]" :

  X = major version  - the overall development status
  Y = minor version  - represents a set of planned features (see TODO);
		       an even number means stable while odd marks 
		       development releases
  Z = revision       - small changes to major.minor version (optional)
		
  devT = dev-release - development releases leading to the next stable
		       version

-----------------------------------------------------------------------

Last officially released version is -*> v2.0.3 *-

????2001 (v2.1dev4)
	- added: eiam - 'cs-option' commmand; although explicit use 
		 of cmd-line options as iactive-mod commands (like 
                 "-i:file.wav") is still supported, "cs-option 
                 -i:file.wav" is now the preferred syntax
	- removed: ecasound FAQ file - it only contains about 	 
		   a pageful of useful text - I moved that
		   precious piece of content to the user's 
		   guide :)
	- changed: uses of '-i' an '-o' replaced with 'ai-add' 
		   and 'ao-add' in ECI documentation and 
		   code examples in 'ecasound/examples'
	- changed: user's and programmer's guides updated
	- changed: bug-alert; major rewrite of ecasound's 
                   internal object factory code
	- changed: eiam - output of aio-register, cop-register,
		   ladspa-register and preset-register commands
	- fixed: multitrack mode was not properly detected if 
		 buffering mode was selected explicitly (-B)
	- fixed: eiam - minor 'cs-edit' bugs fixed
	- fixed: 'make install' failed if srcdir != objdir
	- fixed: didn't work with ALSA +0.9b8 CVS-tree
21102001 (v2.1dev3)
	- note: major changes to effect preset system which 
		break some old preset definitions
	- edi-entries: edi-3
	- added: ecasound-config - --ldflags and --prefix
	- added: engine profiling info (printed to stderr)
	- added: -z:multitrack, -z:nomultitrack for forced enabling
		 and disabling of multitrack-mode
	- added: variable-like preset parameters; it's now possible
		 to bind one preset parameter to multiple chainops'
		 parameters (control multiple settings with one
		 parameter)
	- added: support for describing preset parameters (edi-3)
	- changed: pyecasound - makefile-hack that allows pyecasound
		   to work even though compiled against libecasound 
		   and libkvutils that are not installed to 
		   a directory listed in /etc/ld.so.conf
	- changed: db-subsystem - is put to sleep more 
		   aggressively, which should further decrease 
		   its cpu-load
	- changed: db-subsystem - runs once again under sched_fifo
		   scheduling if -r is enabled
	- changed: ALSA xrun reports are now printed to stderr 
		   instead of ecasound debug-subsystem
	- changed: -B:rt selected as the default bmode for 
		   multitrack-setups (instead of -B:lowlatency)
	- changed: preset definitions in 'effect_presets' changed
		   to take advantage of the preset system improvements
	- fixed: major bugs in interpreting LADSPA plugin 
		 parameter hint flags
	- fixed: a nasty bug in bufmode activation code prevented
		 -kos, -kl, and others from working properly
	- fixed: loop devices didn't work well with -z:db; fixes
		 ecasound testsuite case con_test1-8
	- fixed: memory leak in LADSPA handling code; was triggered
	 	 by recent versions of ecamegapedal
	- fixed: pyecasound - make target 'install-exec-hook' failed
		 if --disable-pyecasound was given to configure
14102001 (v2.1dev2)
	- added: ecaplay - -k, -d and -h options, see 'man ecaplay'
	- added: eiam - ai-get-length-samples, ao-get-length-samples,
		 cs-get-length-samples, cs-set-length-samples [smassy]
	- added: support for variable length buffer's for ALSA 0.9
		 output (required by -ei:xx type effects)
	- changed: tuned prefill-parameters for double-buffering
	- changed: when saving chainsetups, if no parameters 
	           are overridden, buffermode definition (-B) is
		   saved instead of individual parameters
	- changed: yet more changes to the multitrack-mode
	- changed: eiam - output of *-status commands
	- fixed: -r option wasn't saved to chainsetups (used
		 to be a session-level option)
	- fixed: typeselect problems when used for outputs
	- fixed: system lockups with -B:rtlowlatency and 
		 rt-object-only chainsetups
	- fixed: misplaced tab prevented eci.py from working
		 with python 2.1 (nested scopes)
	- fixed: if a monitor track input ended before recording 
                 stopped in multitrack mode and -z:db was enabled, 
		 it continued producing unwanted noise to the
		 monitor output
	- fixed: eiam - ladspa-register output had plugin names 
		 and descriptions mixed
09102001 (v2.1dev1)
	- added: chainsetup parser prints a warning if an unknown
	         buffering mode is given to -B:mode 
	- changed: 'rt' buffering mode is preferred over 
		   'rtlowlatency' if no chain operators are 
		   present, and multitrack mode is not active
        - changed: more improvements to double-buffering
	- fixed: critical bug when running in -B:rtlowlatency 
		 mode (caused system-lockups with ALSA 0.9.x
		 oss-emulation)
	- fixed: explicitly specifying -z:nointbuf had no 
		 effect
	- fixed: setting double-buffer length didn't have 
		 any effect (given either in bmode defaults 
		 or as -z:db,dbsize)
06102001 (v2.1dev0)
	- edi-entries: edi-2, edi-6, edi-10, edi-15
	- added: more recent config.guess and config.sub to 
	         the dist package  [juekawa]
	- added: new feedback-% param to -etd effect, doesn't
		 affect old presets using -etd [smassy]
	- added: ALSA device name 'default' is used if no 
                 other name is given ('-o:alsa' is equivalent
	         to '-o:alsa,default')
	- added: special case for mapping "ecasound file.ecs"
		 to "ecasound -s:file.ecs"
	- added: implementation of edi-6; the -z:db

ecatools/ecalength.c
/*   
    This is ecalength, a few lines of code pompously named so because they  
    let one retrieve the length of an audio file from the command line  
    using ecasound's engine.  

    Limitations:  
    - With files without header information (raw files), ecalength will only work 
      correctly if the audio file is at a sampling rate of 44100 hz.
      (Addressed with the -a switch.)
    - It is not foolproof, feeding it with something other than an audio  
      file WILL result in ugly things being spewed back.  
      (A bit better)
    - A thousand more that I haven't thought of.  

    Please post back any improvement you make; I can be reached at:  
    observer@colba.net  

    note: Compile it with:  
    gcc -Wall -lecasoundc -o ecalength ecalength.c  

    last updated: Thu May 10 15:56:18 EDT 2001
- Now works with the new ai/ao scheme.
- Switches implemented, made suitable for scripting.
- Format querying/setting.
- Better error handling.  
*/ 

#include <stdio.h> 
#include <stdlib.h> 
#include <unistd.h> 
#include <string.h>
#include "ecasoundc.h" 

#define FALSE          0 
#define TRUE           1 

void make_human(int length, unsigned int *min, unsigned char *sec); 
void print_usage(char* name); 

struct options { 
  char adjust;
  char format; 
  char total; 
  char script; 
  char human;
  char bits;
  char ccount; 
  char rate;
}; 

int main(int argc, char *argv[]) { 
  char cmd[500], fstring[16], status = 0, curopt, *optstr = "ftsmhbcra:"; 
  unsigned char sec; 
  float curfilelength, totlength = 0; 
  unsigned int min, curarg; 
  FILE *file; 
  struct options opts; 

  opts.adjust = FALSE;
  opts.format = FALSE; 
  opts.total = FALSE; 
  opts.script = FALSE; 
  opts.human = FALSE; 
  opts.bits = FALSE;
  opts.ccount = FALSE;
  opts.rate = FALSE;

  while ((curopt = getopt(argc, argv, optstr)) != -1) { 
    switch (curopt) { 
    case 'a' : opts.adjust = TRUE;
      strcpy(fstring, optarg);
      break;
    case 'f' : opts.format = TRUE; 
      break; 
    case 't' : opts.total = TRUE; 
      break; 
    case 's' : opts.script = TRUE; 
      break; 
    case 'm' : opts.human = TRUE; 
      break; 
    case 'b' : opts.bits = TRUE;
      break;
    case 'c' : opts.ccount = TRUE;
      break;
    case 'r' : opts.rate = TRUE;
      break;
    case 'h' : print_usage(argv[0]);
      exit(0);
    case '?' : print_usage(argv[0]);
      exit(1);
    } 
  } 

  if (argc-optind == 0) {
    print_usage(argv[0]);
    exit(1);
  }

  if ((opts.script) && (((opts.format) && (opts.human)) ||
			((opts.format) && (((opts.bits) && ((opts.ccount) || 
							    (opts.rate))) ||
					   ((opts.ccount) && (opts.rate)))))) {
    fprintf(stderr, "Error: In script mode not more than one further mode can be specified.\n");
    print_usage(argv[0]);
    exit(1);
  }

  eci_init(); 
  eci_command("cs-add main"); 
  eci_command("c-add main"); 
  eci_command("ao-add null"); 
  if (opts.adjust) {
    if (strncmp(":", fstring, 1) == 0) { sprintf(cmd, "cs-set-audio-format %s", fstring+1); }
    else { sprintf(cmd, "cs-set-audio-format %s", fstring); }
    eci_command(cmd);
    if (strlen(eci_last_error()) != 0) {
      fprintf(stderr, "Argument to -a is badly formatted.\n");
      print_usage(argv[0]);
      exit(1);
    }
  }

  curarg = optind; 

  while(curarg < argc) { 
    if ((file = fopen(argv[curarg], "r")) != NULL) { 
      fclose(file); 
      sprintf(cmd, "ai-add %s", argv[curarg]); 
      eci_command(cmd); 
      eci_command("cs-connect"); 
      if (strlen(eci_last_error()) == 0) {
	sprintf(cmd, "ai-select %s", argv[curarg]); 
	eci_command(cmd); 
	eci_command("ai-get-length"); 
	curfilelength = eci_last_float(); 
	if (opts.format) { 
	  eci_command("ai-get-format"); 
	  strcpy(fstring, eci_last_string()); 
	} 
	eci_command("cs-disconnect"); 
	eci_command("ai-remove"); 
	if (!(opts.script) || ((opts.script && opts.human))) { 
	  make_human((int)(curfilelength+0.5), &min, &sec); 
	} 
	if (!(opts.script)) { printf("%s: ", argv[curarg]); } 
	if (!(opts.script) ||  
	    ((opts.script) && (!(opts.format) && !(opts.human)))) { 
	  printf("%.3f", curfilelength); 
	} 
	if (!(opts.script)) { printf("s   \t("); } 
	if (!(opts.script) || ((opts.script) && (opts.human))) { 
	  printf("%im%is", min, sec); 
	} 
	if (!(opts.script)) { printf(")"); } 
	if ((opts.format) && 
	    !((opts.format) && ((opts.bits) || (opts.ccount) || (opts.rate)))) { 
	  if (!(opts.script)) { printf("   \t"); } 
	  printf("%s", fstring); 
	} 

	if ((opts.format) && (opts.script) && (opts.bits)) { 
	  printf("%s", strtok(fstring+1, "_")); 
	}

	if ((opts.script) && (opts.format) && (opts.ccount)) {
	  strtok(fstring, ",");
	  printf("%s", strtok(NULL, ","));
	}

	if ((opts.format) && (opts.script) && (opts.rate)) {
	  strtok(fstring, ",");
	  strtok(NULL, ",");
	  printf("%s", strtok(NULL, ","));
	}

	printf("\n"); 
	if ((opts.total) && !(opts.script)) { 
	  totlength += curfilelength; 
	} 
      }
	  else {
	    if (opts.script) { printf("-2\n"); }
	    else { printf("%s: Read error.\n", argv[curarg]); }
	    status = -2;
	    eci_command("ai-remove");
	  }
    } 
    else { 
      if (opts.script) { printf("-1\n"); }
      else { printf("%s: fopen error.\n", argv[curarg]); }
      status = -1;
    } 
    curarg++; 
  } 

  if ((opts.total) && !(opts.script)) { 
    make_human((int)(totlength+0.5), &min, &sec); 
    printf("Total: %.3fs \t\t(%im%is)\n", totlength, min, sec); 
  } 

  eci_command("cs-remove");
  eci_cleanup();

  exit(status); 
} 

void make_human(int length, unsigned int *min, unsigned char *sec) { 
  *min = (length/60); 
  *sec = (length % 60); 
} 

void print_usage(char *name) { 
  printf("Usage: %s [-ahtsfmbcr] FILE1 [FILE2] [FILEn]\n 
       -h      Prints this usage message.  (help)
       -a[:]bits,channels,rate     Changes the format assumed by default 
                                   for headerless data.  (adjust)
       -t      Prints the summed length of all the files processed.  (total)
                 (Ignored if with -s) 
       -s      Enables script mode: One info type per file per line.   (script)
                 (Defaults to length in secs.) 
       -f      With -s will return the format string as info, alone it will 
               add it to the main display.  (format)
           -b  If -s and -f are enabled with this the info printed will be 
              the sample's bitwidth.  (bits)
           -c  If -s and -f are enabled with this the info printed will be 
               the channel count.  (channel count)
           -r  If -s and -f are enabled with this the info printed will be 
               the sampling rate.  (rate)
       -m      Will print human computable time as in main display but in 
               batch fashion.  (minutes)
                 (Only with -s)
   (Note that out of context options will be silently ignored.)\n\n",
	 name);
}
